src/scala/microsoft-spark-2.4.x/src/main/scala/org/apache/spark/api/dotnet/CallbackClient.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the .NET Foundation under one or more agreements.
+ * The .NET Foundation licenses this file to you under the MIT license.
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.apache.spark.api.dotnet
+
+import java.io.{DataInputStream, DataOutputStream}
+
+import org.apache.spark.internal.Logging
+
+import scala.collection.mutable.Queue
+
+/**
+ * CallbackClient is used to communicate with the Dotnet CallbackServer.
+ * The client manages and maintains a pool of open CallbackConnections.
+ * Any callback request is delegated to a new CallbackConnection or
+ * unused CallbackConnection.
+ * @param address The address of the Dotnet CallbackServer
+ * @param port The port of the Dotnet CallbackServer
+ */
+class CallbackClient(address: String, port: Int) extends Logging {
+  private[this] val connectionPool = Queue[CallbackConnection]()
+
+  private[this] var isShutdown = false
+
+  final def send[T](writeBody: DataOutputStream => Unit,
+                    readBody: Option[DataInputStream => T],
```

Comment:
Where do you set non-None readBody? (I may have missed it)

Reply:
This was set in `DotnetForeachBatchFunction.call`.  I'll put it in a newline so it's easier to see.

Reply:
yea, but it was `None`. Any `Some` case?

Reply:
No Some case since ForeachBatch callback doesn't expect a return value. This is here in case we need a callback to return something, maybe in the future.

Reply:
Can we remove the code then? Unless you already have a test case for this.

Reply:
I have test case in the dotnet callbackserver/connection.  Only test I did on scala was an ad-hoc one by modifying DotnetForeachBatchFunction.call to add a writeBody and retrieving the result.

Reply:
is it possible to support `void` with this code path then? we already have too many ad-hoc testing (e.g., retry, error handling, etc.). Since we don't have a testing framework on Scala side, I want to minimize the changes on the Scala side as much as possible.

Reply:
You mean have 2 public send methods?

```scala
  final def send[T](
      callbackId: Int,
      writeBody: DataOutputStream => Unit,
      retries: Int = 3): Unit = {
    send(callbackId, writeBody, None, retries)
  }

  final def send[T](
      callbackId: Int,
      writeBody: DataOutputStream => Unit,
      readBody: DataInputStream => T,
      retries: Int = 3): T = {
    val resultOption = send(callbackId, writeBody, Some(readBody), retries)
    resultOption match {
      case Some(result) => result
      case None => throw new Exception("Return value expected, but none received.")
    }
  }

  private final def send[T](
      callbackId: Int,
      writeBody: DataOutputStream => Unit,
      readBody: Option[DataInputStream => T],
      retries: Int = 3): Option[T] = {
...
  }
```

Reply:
Nm..  I just removed the support for return values.